pageserver/src/restore_local_repo.rs

```diff
@@ -0,0 +1,460 @@
+        if let Err(e) = open_result {
+            if e.kind() == std::io::ErrorKind::NotFound {
+                break;
+            }
+            return Err(e)?;
+        }
+        let mut file = open_result.unwrap();
+
+        if offset > 0 {
+            file.seek(SeekFrom::Start(offset as u64))?;
+        }
+
+        let mut buf = Vec::new();
+        let nread = file.read_to_end(&mut buf)?;
+        if nread != 16 * 1024 * 1024 - offset as usize {
+            // Maybe allow this for .partial files?
+            error!("read only {} bytes from WAL file", nread);
+        }
+        waldecoder.feed_bytes(&buf);
+
+        let mut nrecords = 0;
+        loop {
+            let rec = waldecoder.poll_decode();
+            if rec.is_err() {
+                // Assume that an error means we've reached the end of
+                // a partial WAL record. So that's ok.
+                break;
+            }
+            if let Some((lsn, recdata)) = rec.unwrap() {
+                let decoded =
+                    crate::waldecoder::decode_wal_record(recdata.clone());
+
+                // Put the WAL record to the page cache. We make a separate copy of
+                // it for every block it modifies. (The actual WAL record is kept in
+                // a Bytes, which uses a reference counter for the underlying buffer,
+                // so having multiple copies of it doesn't cost that much)
+                for blk in decoded.blocks.iter() {
+                    let tag = BufferTag {
+                        spcnode: blk.rnode_spcnode,
+                        dbnode: blk.rnode_dbnode,
+                        relnode: blk.rnode_relnode,
+                        forknum: blk.forknum as u8,
+                        blknum: blk.blkno,
+                    };
+
+                    let rec = page_cache::WALRecord {
+                        lsn: lsn,
+                        will_init: blk.will_init || blk.apply_image,
+                        rec: recdata.clone(),
+                    };
+
+                    pcache.put_wal_record(tag, rec);
+                }
+
+                // Now that this record has been handled, let the page cache know that
+                // it is up-to-date to this LSN
+                pcache.advance_last_valid_lsn(lsn);
+                last_lsn = lsn;
+            } else {
+                break;
+            }
+            nrecords += 1;
+        }
+
+        info!("restored {} records from WAL file {}", nrecords, filename);
+
+        segno += 1;
+        offset = 0;
+    }
+    info!("reached end of WAL at {:X}/{:X}", last_lsn >> 32, last_lsn & 0xffffffff);
+
+    Ok(())
+}
+
+// FIXME: copied from xlog_utils.rs
+pub const XLOG_FNAME_LEN: usize = 24;
+pub type XLogRecPtr = u64;
+pub type XLogSegNo = u64;
+pub type TimeLineID = u32;
+
+#[allow(non_snake_case)]
+pub fn XLogSegmentOffset(xlogptr: XLogRecPtr, wal_segsz_bytes: usize) -> u32 {
+    return (xlogptr as u32) & (wal_segsz_bytes as u32 - 1);
+}
```

Comment:
Why the non_snake_case names?

Reply:
These functions are reimplementations of PostgreSQL functions/macros with same names.

Reply:
Ah. I'd vote to snake_case them anyway, but I have been the style police enough for this week, so I'll leave it up to you :)

Reply:
> Ah. I'd vote to snake_case them anyway,

yeah, I could go either way. We could also name them along the lines of `pg_XLogSegmentOffset`. Although that kind of prefixing isn't very idiomatic in Rust either, because you can use the module name as part of the calls instead, like `xlog_utils::XLogSegmentOffset`. In https://github.com/zenithdb/zenith/pull/50, I'm moving this so that it becomes `postgres_ffi::xlog_utils::XLogSegmentOffset`. I hope the `postgres_ffi` prefix makes it less surprising, the 'ffi' is a hint that there is foreign stuff in it. Then again, that's really long, so we're probably going to use `use` and not spell out the whole module path.

> but I have been the style police enough for this week, so I'll leave it up to you :)

Hey, it's only Tuesday, not casual Friday yet :-).

Reply:
I was going to suggest that some of these might move inside the types they operate on, e.g.
```
impl XLogSegment {
    pub fn get_offset(lsn: Lsn) -> u32 { ... }
}
```